stan/math/opencl/matrix_cl.hpp

```diff
@@ -294,12 +294,48 @@ class matrix_cl<T, require_arithmetic_t<T>> {
    * @throw <code>std::system_error</code> if the memory on the device could not
    * be allocated
    */
-  template <typename Mat, require_eigen_t<Mat>..., require_same_vt<Mat, T>...>
+  template <typename Mat, require_eigen_t<Mat>..., require_same_vt<Mat, T>...,
+            require_not_t<is_eigen_contiguous_map<Mat>>...>
   explicit matrix_cl(Mat&& A,
                      matrix_cl_view partial_view = matrix_cl_view::Entire)
       : rows_(A.rows()), cols_(A.cols()), view_(partial_view) {
-    initialize_buffer<std::is_rvalue_reference<Mat&&>::value
-                      || !is_eigen_matrix<Mat>::value>(A.eval().data());
+    using Mat_type = std::decay_t<decltype(A.eval())>;
+    if (size() == 0) {
+      return;
+    }
+    if (is_eigen_matrix_or_array<Mat>::value
+        && std::is_lvalue_reference<Mat>::value) {
+      initialize_buffer(A.eval().data());
+    } else {
+      auto* A_heap = new Mat_type(std::move(A.eval()));
+      try {
+        cl::Event e = initialize_buffer(A_heap->data());
```

Comment:
So first off, this is a side-effect. `initialize_buffer` now returns a `cl::Event` and does not return `void`. Please document this in the PR and help explain the design.

This is definitely leak-able. You've handled L312 appropriately: the `initialize_buffer()` call can throw. There's a catch which deletes the memory.

L313 can definitely leak. If the [sdk documentation](https://www.khronos.org/registry/OpenCL/sdk/2.0/docs/man/xhtml/clSetEventCallback.html) for `clSetEventCallback()` is correct, then we must check the error code returned to determine if there has been an error. If any of the errors described in that doc happens, then `e` is still constructed how it's constructed, but the callback is not registered and this memory will never be freed. I see that you've repeated this pattern multiple times, so it'll have to be fixed in each place.

Reply:
> So first off, this is a side-effect.

Can you explain what do you mean with that?

> initialize_buffer now returns a cl::Event and does not return void. Please document this in the PR and help explain the design.

`initialize_buffer` return value is documented in doxygen. Or did you mean I should more deeply explain here? We now in some cases (the one I am changing) need the event for the copy outside of the `initialize_buffer` function. So the return. Any other use can still ignore the return as if ti was void.

I am not calling `clSetEventCallback()` directely. We use c++ bindings which support throwing exceptions on error conditions if `CL_HPP_ENABLE_EXCEPTIONS` macro is defined. And we do define it. So `e.setCallback()` throws if `clSetEventCallback()` would return error value. So I think there are no bugs.

Reply:
> Can you explain what do you mean with that?

Sorry about being imprecise with terminology. I was actually mixing two meanings of the term "side effect;" the first is what's used commonly in [computer science (link to wikipedia)](https://en.wikipedia.org/wiki/Side_effect_(computer_science)). This change really modifies states outside of `matrix_cl`, so it is a side effect and it should really be documented at the PR level (since this is the PR that introduces it) and also in the class itself. The existence of side effects aren't necessarily bad or good, but documenting them is something we must do. It's just a precondition for having any PR merged.

The second use of side effect was that this changes the API. It's interesting because it used to be listed in the same wikipedia article as "In computer science, a function or expression is said to have a side effect if it modifies some state or has an observable interaction with calling functions or the outside world," but was changed recently. It's still [quoted on StackOverflow](https://softwareengineering.stackexchange.com/a/325520).

> `initialize_buffer` return value is documented in doxygen. Or did you mean I should more deeply explain here? We now in some cases (the one I am changing) need the event for the copy outside of the `initialize_buffer` function. So the return. Any other use can still ignore the return as if ti was void.

Yes, now that you're introducing branching logic, this has to be documented for `initialize_buffer` -- for more concreteness, please leave enough documentation for a different person to be able to understand what cases there are, under what conditions they are responsible for dealing with things, and under what conditions they are not supposed to do things. 

> I am not calling clSetEventCallback() directely. We use c++ bindings which support throwing exceptions on error conditions if CL_HPP_ENABLE_EXCEPTIONS macro is defined. And we do define it. So e.setCallback() throws if clSetEventCallback() would return error value. So I think there are no bugs.

Thanks for the explanation! First, I'll start with questions regarding the implementation and leaking conditions:
1. Is the callback guaranteed to be called on normal exit conditions? If so, mind pointing out the documentation that explicitly writes this out?
2. Is the callback guaranteed to be called on abnormal exit conditions? If not, what happens? Can you point out the documentation that states this?
3. Is the callback guaranteed to be called exactly once? More specifically, are there any conditions under which this could be called two or more times? In particular, is any of this done in parallel where there could be a race condition where the callback could be triggered twice due to asynchronous processes trying to figure out ending conditions?
4. Is it possible for the event to continue to be used after the callback?
5. When is the event supposed to be cleaned up? (Since it's outside of this function, just a quick description would help.)

Since you're changing this, I think we should talk about design; this PR skipped that phase and went directly into implementation. @SteveBronder brought up a good point about RAII and I think it's appropriate. With this sort of memory-management, it would make sense to encapsulate this logic. It could even be done using some sort of factory function. (There's no reason `matrix_cl` needs to allocate memory and conditionally delete it... we could have a factory function that creates `cl::Event` that either allocates memory and adds the callback or throws and leaves the world in a recoverable place.)

Reply:
Dan I think before we go into the above the base thing I am confused about is what it's solving. Is this to handle the async data movement? If that's the case, can you write a test to show what behavior this fixes?

Reply:
So are side effect in the PR template meant as a changes that are not required to fix an issue or any side effects of functions a PR touches?

I allways assumed it was the first case. If it is the second one I think PR template should be changed to "Side effects in functions touched by this PR" or something similar. 

Also starting any async kernel launch or data transfer is essentially a side effect of a function. I think it was never mentioned as a side effect of a PR that introduced it. So I guess the first case is the right one?

> Yes, now that you're introducing branching logic, this has to be documented for initialize_buffer

That is not correct. I am adding no new branches to `initialize_buffer` in this PR.

>Thanks for the explanation! First, I'll start with questions regarding the implementation and leaking conditions:
Is the callback guaranteed to be called on normal exit conditions? If so, mind pointing out the documentation that explicitly writes this out?
Is the callback guaranteed to be called on abnormal exit conditions? If not, what happens? Can you point out the documentation that states this?

Questions 1 and 2 are irrelevant for this PR - all allocated memory is freed anyway on program exit regardless of whether it is normal or abnormal.

>Is the callback guaranteed to be called exactly once? More specifically, are there any conditions under which this could be called two or more times? In particular, is any of this done in parallel where there could be a race condition where the callback could be triggered twice due to asynchronous processes trying to figure out ending conditions?

Yes it will always be called exactley once. OpenCL API is thread-safe.

> Is it possible for the event to continue to be used after the callback?

Yes. Callback does not affect the event in any way.

> When is the event supposed to be cleaned up?

`cl::Event` object is within the function `initialize_buffer()`. So it is destroyed when the function exits. The event this object contains handle to is a part of OpenCL implementation. User of OpenCL is not required to manage memory of OpenCL implementation's internal objects.

Reply:
>I am confused about is what it's solving. Is this to handle the async data movement?

Yes. With your async PR #1162 we started using asynchronous versions of all OpenCL calls. Than in PR #1356 the bug with constructing `matrix_cl` from rvalues was fixed in simplest possible way. Which meant that when `matrix_cl` is constructed from rvalue blocking call is used. Here I am reverting that change from PR #1356 meaning that async calls can also be used in these cases.

> If that's the case, can you write a test to show what behavior this fixes?

No. Even if we use async calls OpenCL implementation is not actually required to copy the data asynchronously. Even if it was, I think we could not reliably test that function returns before copying is complete. We also don't have any tests that check whether other OpenCL calls are actually done asynchronously.

Reply:
> > I am confused about is what it's solving. Is this to handle the async data movement?

> Yes. 

Can you change the name of this PR then? It sounds like this is a bugfix and not an optimization. It may be good to explain some of the idea around the bugfix as well in the top level comment

> With your async PR #1162 we started using asynchronous versions of all OpenCL calls. Than in PR #1356 the bug with constructing matrix_cl from rvalues was fixed in simplest possible way. Which meant that when matrix_cl is constructed from rvalue blocking call is used. Here I am reverting that change from PR #1356 meaning that async calls can also be used in these cases.

>> If that's the case, can you write a test to show what behavior this fixes?

> No. Even if we use async calls OpenCL implementation is not actually required to copy the data asynchronously. Even if it was, I think we could not reliably test that function returns before copying is complete. We also don't have any tests that check whether other OpenCL calls are actually done asynchronously.

I'm very wary about fixing bugs we can't replicate on at least a particular setup. If we need a particular environment for a specific bug we may be able to make a dockerfile to replicate that environment. The platform and device both have to have async support, we can test whether we have that and if not then we can probably make an ec2 setup that has it. i.e.  doing the test on a new intel cpu with their platform since they support async.

Is there a reason something like below wouldn't work

```cpp
TEST(yadayada, blah) {
stan::math::matrix_cl;
for (i in 1:ALot) {
// do a bunch of opencl moving around and ops to build up the async stack
}

matrix_cl foo(Eigen::BigIdentityMatrix(BigN) + 1);
int check_val = func_to_flag_if_foo_not_right(foo);
EXPECT_TRUE({check_val == 1})

```


>> Is the callback guaranteed to be called on normal exit conditions? If so, mind pointing out the documentation that explicitly writes this out?
Is the callback guaranteed to be called on abnormal exit conditions? If not, what happens? Can you point out the documentation that states this?

> Questions 1 and 2 are irrelevant for this PR - all allocated memory is freed anyway on program exit regardless of whether it is normal or abnormal.

They are relevant. Page 179 of the [OpenCL Spec](https://www.khronos.org/registry/OpenCL/specs/opencl-1.2.pdf) gives the end conditions of an event. The setup you have here

```cpp
        e.setCallback(CL_COMPLETE, &delete_it<U_val>, obj_heap);
      } catch (...) {
        delete obj_heap;
        throw;
      }
```

Will delete `obj_heap` when the callback is complete or throws when inside of this functions scope, but it won't delete `obj_heap` if the event returns an error code when the program has exited this scope. Though I'm pretty sure since we set OpenCL up so that it terminates the program when there is a hardware error we would just clean up the whole program. That's a side effect though directly from the code in this PR. I've thought about turning those failures into domain errors instead of terminal conditions. For example when we start having multiple chains in a stan program it very well may happen that the chains send too much stuff over sometimes and we run out of gpu ram. But that's not guaranteed to fail on every iteration so we may just want to throw a `domain_error` for that and let the program keep running. 

For Q 2 this paragraph seems to imply that abnormal exit conditions in parts of the program before a call is finished is implementation defined

> If the execution of a command is terminated, the command-queue associated with this
terminated command, and the associated context (and all other command-queues in this context) may no longer be available. The behavior of OpenCL API calls that use this context (and command-queues associated with this context) are now considered to be implementation defined. The user registered callback function specified when context is created can be used to report appropriate error information.

There's also some stuff on 186 about how expensive ops can also lead to undefined behavior. I actually have no idea if deletion of a large object is considered expensive or not in the OpenCL Spec's idea of 'expensive'.

> So are side effect in the PR template meant as a changes that are not required to fix an issue or any side effects of functions a PR touches?
> I always assumed it was the first case. If it is the second one I think PR template should be changed to "Side effects in functions touched by this PR" or something similar.

It would be nice to clear up the PR templates 'side effects' section. I never know what to put there. I certainly hope my PRs don't have unintended side effects! Should it be 'consequences' for other parts of the stan math library or something like that?

I'm not sure if "side effect" is the right word for that part of the PR template, at least it has a lot of definitions. Though this is probs a conversation for somewhere else.

Reply:
> It sounds like this is a bugfix and not an optimization.

I am not sure if we call something potentially working slower a bug. Whatever, it is a matter of definition, so I don't mind changing title if you are sure it should be done.

> Is there a reason something like below wouldn't work

I think it will work on some setups, but definitely not all of them.

> They are relevant.

Oh, we are talking about kernel execution status? I understood exit conditions as what is happening at program exit. The doc [here](https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/clSetEventCallback.html) states: "The callback function registered for a command_exec_callback_type value of CL_COMPLETE will be called when the command has completed successfully or is abnormally terminated. " So the answer to both questions is yes, it is guaranteed.

> For Q 2 this paragraph seems to imply that abnormal exit conditions in parts of the program before a call is finished is implementation defined

> the behavior of OpenCL API calls that use this context (and command-queues associated with this context) are now considered to be implementation defined. The user registered callback function specified when context is created can be used to report appropriate error information.

It only says API calls that reference context are undefined. What we use is exactley what last sentence mentiones - a callback that was registered when the context was still valid to get some info on error.

> There's also some stuff on 186 about how expensive ops can also lead to undefined behavior. I actually have no idea if deletion of a large object is considered expensive or not in the OpenCL Spec's idea of 'expensive'.

That part seems a bit weird. I dont understand how a function that takes more than certain amount of time could break anything. Anyway I found [another OpenCL function](https://www.khronos.org/registry/OpenCL/sdk/2.0/docs/man/xhtml/clSetMemObjectDestructorCallback.html) with the same warning in doc that also states that its typical use is freeing the memory in the callback. So I guess this should be fine.

Reply:
@t4c1: thank you for the responses! Especially providing the relevant parts of the API spec as links. For future reference, it would help if you, as the submitter, provided those references for 3rd party libraries. It’s hard to expect a reviewer to know any of the libraries well enough to know the edge cases. Since you’ve linked them in the discussion, can you update the doxygen documentation with this info? It should have been there with the PR as part of what is necessary to keep this maintainable. Thank you for pointing out the documentation. I now understand why you coded it this way and believe that the callback should be called once and only once and at the end. 

Re: branching. I think you’re misunderstanding the scope. This PR clearly introduces branching. Starting with L306:
```
	    if (is_eigen_matrix_or_array<Mat>::value
        && std::is_lvalue_reference<Mat>::value) {
      initialize_buffer(A.eval().data());
    } else {
      auto* A_heap = new Mat_type(std::move(A.eval()));
      try {
        cl::Event e = initialize_buffer(A_heap->data());
...
```

Prior to this PR, there was no if/else.

Re: expensive ops. This is where we have to code defensively. We should be coding defensively all the time, but APIs change. If we’re relying on certain behavior in order for our functions to work, we should have tests that check that condition (not necessarily as part of the code, but it could be a unit test that fails).

Question: is there any way that an event could fail to trigger the callback, but have this continue?

Also, why isn’t the new function `initialize_buffer_optionally_from_heap` used? It seems like this is the behavior I was describing. There’s no need for this constructor to call `new` and `delete` on its own. 

Also, have you considered using the memory arena that’s part of the autodiff stack? If it’s there, it won’t leak because the memory locations aren’t allocated when called and it is the burden of the client to recover that memory every time. 

@SteveBronder: thank you! The review comments you’re making are spot on, especially asking about the high-level context of the PR. 

Re: PR template and “side effect.” It was meant as everything. It was really meant for the submitter to write what they thought were side effects, either in code, changes to the api, changes to functions, changes to design. It’s really hard to be rigid about this because any time we do anything interesting, it’ll fall outside what we write. And if we have narrowly scoped things, most of the time it won’t be used. But.... if what’s written is not useful, then we should update it. Ideally, any new side effects in functions would be documented in the doxygen doc and then highlighted in the PR so the reviewer is told what side effects there are. That makes it easy for the reviewer. But that’s not the only way it could be done. Anyway, if you can make that clearer, please do.

Reply:
> It’s hard to expect a reviewer to know any of the libraries well enough to know the edge cases.

Well its only natural that there are questions and possibly some misunderstandings when looking at some code for the first time. One of the reasons for reviewing is that those are addressed. BTW next time you need some documentation for OpenCL you can google "OpenCL [function name]" and the first result should be the doc.

> Yes, now that you're introducing branching logic, this has to be documented for initialize_buffer
> I think you’re misunderstanding the scope. 
That seems to be the case. Since you mention `initialize_buffer()` in the same sentence I naturally assumed you are talking about branching in this function. 

> Re: expensive ops. This is where we have to code defensively. We should be coding defensively all the time, but APIs change. If we’re relying on certain behavior in order for our functions to work, we should have tests that check that condition (not necessarily as part of the code, but it could be a unit test that fails).

If something crashes or deadlocks any test will detect it. If undefined behavior is something else we may or may not detect it. But as I explained in previous post I am pretty sure we are not actually triggering any undefined behavior.

> Question: is there any way that an event could fail to trigger the callback, but have this continue?

I think the only way for this to happen is that program exits before copying is completed.

> Also, why isn’t the new function `initialize_buffer_optionally_from_heap` used?

It does not do the same thing! `initialize_buffer_optionally_from_heap()` copies an object to heap. Which would be fine for matrices, but not eigen expressions. These need to be evaluated before they can be copied to GPU. In other words the difference is in `.eval()` (which is a no-op for matrices).

> Also, have you considered using the memory arena that’s part of the autodiff stack?

No. It in an interestiong idea, but I have some concerns:
1. Do wa have a way to guarantee that memory is not released before copying to GPU is complete?
2. As far as I understand it one of advantages of using arena is memory locality for autodiff variables that are often used together. Adding matrices that are only used once in copying would decrease that locality.

I am moving the discussion about side effects to discourse: https://discourse.mc-stan.org/t/what-are-side-effects-in-pr-template/11776

Reply:
So... I think we're converging on the questions of the lifetime of the `matrix_cl` object, the `cl::Event` object, and the lifetime of the memory in the autodiff stack. That's great!

It wasn't really described in this PR... how long does the `cl::Event` object live? Will `matrix_cl` be in scope longer than the `cl::Event` object that's created? Under what conditions will the `cl::Event` object still be active after `matrix_cl` falls out of scope? Will it ever continue past the computation of a gradient?

We do have arena memory already allocated that's not used for autodiff. It could possibly used for this application. We do that in other parts of the Math library for different things, so it's not a new pattern. The caller is responsible for calling `recover_memory()` after running the reverse sweep of autodiff (calling the `grad()` function or `.chain()` method). I believe that should be enough time. Since it's on the caller, there's no guarantee that the caller will not call `recover_memory()` while building up the stack, but the Math library treats that as undefined behavior. I think from the outside, a client would call `recover_memory()` on an exception, so that might be the condition we have to think through... if we kicked of something with the `cl::Event` still running, but then had an exception and the memory was recovered, what would happen? 


>  BTW next time you need some documentation for OpenCL you can google "OpenCL [function name]" and the first result should be the doc.

FYI, I did do that and I found that particular doc. It really would have helped anyone reviewing if that was included in the PR as support for the design you've used. That said, it wasn't enough information. It would have just been a really good starting point.

Reply:
> It wasn't really described in this PR... how long does the cl::Event object live? 

`cl::Event` has no different lifetime than any other object - untill the end of the scope it is declared in. In this case that is until the constructor returns. How underlying implementation handles callback so that they can be executed after the event stops existing is not defined.

> Will matrix_cl be in scope longer than the cl::Event object that's created?

It can be and I think it usually would be.

> Under what conditions will the cl::Event object still be active after matrix_cl falls out of scope?

Please be more precise. What active means here? I assume you think in what cases callback is not yet called when matrix goes out of scope. That would be if copying the data is not complete, regardles of how far the execution of the program moved forward.

> but then had an exception and the memory was recovered, what would happen?

It is not defined in the doc so i guess it is undefined behavior. The matrix could contain garbage data or the program would simply segfault.

Thanks for the explanation about arena memory. Unfortunately it raises another concern. That would mean we are essentially leaking data untill the end of the reverse sweep. We could quickly run out of memory if this was used in an iterative algorithm.

Reply:
> Thanks for the explanation about arena memory. Unfortunately it raises another concern. That would mean we are essentially leaking data untill the end of the reverse sweep. We could quickly run out of memory if this was used in an iterative algorithm.

Great point! Yes, I didn't quite understand the use case behind this. So... this could be called multiple times within one reverse sweep? Is it possible for there to be more than one `cl::Event` at a time? If so, is there some limit?

> `cl::Event` has no different lifetime than any other object - untill the end of the scope it is declared in. In this case that is until the constructor returns. How underlying implementation handles callback so that they can be executed after the event stops existing is not defined.

This doesn't make sense to me. This is L311-314:
```
      try {
        cl::Event e = initialize_buffer(A_heap->data());
        e.setCallback(CL_COMPLETE, &delete_it<Mat_type>, A_heap);
      } catch (...) {
```

I think you're saying that the `cl::Event` object itself has a lifetime that lasts only within that try block. If that's the case, then why is `setCallback()` a method on `cl::Event`? This code leaves something behind in some global state and it clearly has some lifetime beyond this try block.

- If the callback is set for the event, `e`, itself, then how does `e` accept the notification for the callback and call the different functions that it has queued up?
- If the callback is set for a global event queue (not for this instance of `e`), then how does the global event queue know how to only call these callbacks for this event (which has fallen out of scope)? If it doesn't distinguish across events and only handles `CL_COMPLETE`, then this would be a problem if more than one of the things that are left behind in the global state and the first one triggered these callbacks... that would clear the memory for all of these things at once.

I think either way, it's pretty confusing. You're right... the object `e` drops out of scope. But something is left behind. That's what really needs to be documented clearly. How / why / when is its lifetime over / what assumptions are we making / etc.

> Please be more precise. What active means here? 

Sorry about that. I was just talking about lifetime. To be more clear, this page on cppreference describes lifetime as I know it: https://en.cppreference.com/w/cpp/language/lifetime.

Here, there are two things: 
1. "For any object of class or aggregate types if it, or any of its subobjects, is initialized by anything other than the trivial default constructor, lifetime begins when initialization ends.
For any object of class types whose destructor is not trivial, lifetime ends when the execution of the destructor begins." -- that's for `cl::Event  e`. 
2. "For all other objects (class objects initialized by a trivial default constructor, non-class objects, arrays of those, etc.), lifetime begins when the properly-aligned storage for the object is allocated and ends when the storage is deallocated or reused by another object." That's for the memory that's been allocated.

If I understand correctly, the `matrix_cl` class passes ownership of the memory allocated on the heap to the newly created `cl::Event e`. There's a callback that's registered with `e`, but `e` falls out of scope immediately, so... who owns the memory at this point?

Reply:
> So... this could be called multiple times within one reverse sweep? Is it possible for there to be more than one cl::Event at a time? If so, is there some limit?

Yes it can. There is no limit to number of the number of times this is called in one sweep and no limit on amount of `cl::Event`s in existance.

> I think you're saying that the cl::Event object itself has a lifetime that lasts only within that try block. If that's the case, then why is setCallback() a method on cl::Event? This code leaves something behind in some global state and it clearly has some lifetime beyond this try block.

That `cl::Event` object only lives until the und of try. OpenCL standard does not define how these callbacks are implemented or what state is left behind in the global state so I can not answer how that and the following quiestions. Also it is entirebly possible that it is not done in same way for multiple OpenCL implementations.

> what assumptions are we making

Only that OpenCL functions do what the standard says they should. Nothing about how they are implemented.

> who owns the memory at this point?

Memory ownership AFAIK does not need to be precisely defined. We could say nobody. Or we could say the `matrix_cl` that scheduled the callback. Or we could say it is OpenCL implementation that is responsible for calling the callback.